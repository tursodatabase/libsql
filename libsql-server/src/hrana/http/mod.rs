use anyhow::{Context, Result};
use bytes::Bytes;
use futures::stream::Stream;
use parking_lot::Mutex;
use serde::{de::DeserializeOwned, Serialize};
use std::pin::Pin;
use std::sync::Arc;
use std::task;

use super::{batch, cursor, Encoding, ProtocolError, Version};
use crate::auth::Authenticated;
use crate::connection::{Connection, MakeConnection};
use crate::hrana::http::stream::StreamError;

mod proto;
mod protobuf;
mod request;
pub(crate) mod stream;

pub struct Server<C> {
    self_url: Option<String>,
    baton_key: [u8; 32],
    stream_state: Mutex<stream::ServerStreamState<C>>,
}

#[derive(Debug, Copy, Clone)]
pub enum Endpoint {
    Pipeline,
    Cursor,
}

impl<C: Connection> Server<C> {
    pub fn new(self_url: Option<String>) -> Self {
        Self {
            self_url,
            baton_key: rand::random(),
            stream_state: Mutex::new(stream::ServerStreamState::new()),
        }
    }

    pub async fn run_expire(&self) {
        stream::run_expire(self).await
    }

    pub async fn handle_request(
        &self,
        connection_maker: Arc<dyn MakeConnection<Connection = C>>,
        auth: Authenticated,
        req: hyper::Request<hyper::Body>,
        endpoint: Endpoint,
        version: Version,
        encoding: Encoding,
    ) -> Result<hyper::Response<hyper::Body>> {
        handle_request(
            self,
            connection_maker,
            auth,
            req,
            endpoint,
            version,
            encoding,
        )
        .await
        .map_err(|e| {
            tracing::error!("hrana server: {}", e);
            e
        })
        .or_else(|err| {
            err.downcast::<StreamError>()
                .map(|err| stream_error_response(err, encoding))
        })
        .or_else(|err| err.downcast::<ProtocolError>().map(protocol_error_response))
    }

    pub(crate) fn stream_state(&self) -> &Mutex<stream::ServerStreamState<C>> {
        &self.stream_state
    }
}

pub(crate) async fn handle_index() -> hyper::Response<hyper::Body> {
    text_response(
        hyper::StatusCode::OK,
        "Hello, this is HTTP API v2 (Hrana over HTTP)".into(),
    )
}

async fn handle_request<C: Connection>(
    server: &Server<C>,
    connection_maker: Arc<dyn MakeConnection<Connection = C>>,
    auth: Authenticated,
    req: hyper::Request<hyper::Body>,
    endpoint: Endpoint,
    version: Version,
    encoding: Encoding,
) -> Result<hyper::Response<hyper::Body>> {
    match endpoint {
        Endpoint::Pipeline => {
            handle_pipeline(server, connection_maker, auth, req, version, encoding).await
        }
        Endpoint::Cursor => {
            handle_cursor(server, connection_maker, auth, req, version, encoding).await
        }
    }
}

async fn handle_pipeline<C: Connection>(
    server: &Server<C>,
    connection_maker: Arc<dyn MakeConnection<Connection = C>>,
    auth: Authenticated,
    req: hyper::Request<hyper::Body>,
    version: Version,
    encoding: Encoding,
) -> Result<hyper::Response<hyper::Body>> {
    let req_body: proto::PipelineReqBody = read_decode_request(req, encoding).await?;
    let mut stream_guard =
        stream::acquire(server, connection_maker, req_body.baton.as_deref()).await?;

    let mut results = Vec::with_capacity(req_body.requests.len());
    for request in req_body.requests.into_iter() {
        tracing::debug!("pipeline:{{ {:?}, {:?} }}", version, request);
        let result = request::handle(&mut stream_guard, auth.clone(), request, version).await?;
        results.push(result);
    }

    let resp_body = proto::PipelineRespBody {
        baton: stream_guard.release(),
        base_url: server.self_url.clone(),
        results,
    };
    Ok(encode_response(hyper::StatusCode::OK, &resp_body, encoding))
}

async fn handle_cursor<C: Connection>(
    server: &Server<C>,
    connection_maker: Arc<dyn MakeConnection<Connection = C>>,
    auth: Authenticated,
    req: hyper::Request<hyper::Body>,
    version: Version,
    encoding: Encoding,
) -> Result<hyper::Response<hyper::Body>> {
    let req_body: proto::CursorReqBody = read_decode_request(req, encoding).await?;
    let stream_guard = stream::acquire(server, connection_maker, req_body.baton.as_deref()).await?;

    let mut join_set = tokio::task::JoinSet::new();
    let mut cursor_hnd = cursor::CursorHandle::spawn(&mut join_set);
    let db = stream_guard.get_db_owned()?;
    let sqls = stream_guard.sqls();
    let pgm = batch::proto_batch_to_program(&req_body.batch, sqls, version)?;
    cursor_hnd.open(db, auth, pgm, req_body.batch.replication_index);

    let resp_body = proto::CursorRespBody {
        baton: stream_guard.release(),
        base_url: server.self_url.clone(),
    };
    let body = hyper::Body::wrap_stream(CursorStream {
        resp_body: Some(resp_body),
        join_set,
        cursor_hnd,
        encoding,
    });
    let content_type = match encoding {
        Encoding::Json => "text/plain",
        Encoding::Protobuf => "application/octet-stream",
    };

    Ok(hyper::Response::builder()
        .status(hyper::StatusCode::OK)
        .header(hyper::http::header::CONTENT_TYPE, content_type)
        .body(body)
        .unwrap())
}

struct CursorStream<D> {
    resp_body: Option<proto::CursorRespBody>,
    join_set: tokio::task::JoinSet<()>,
    cursor_hnd: cursor::CursorHandle<D>,
    encoding: Encoding,
}

impl<D> Stream for CursorStream<D> {
    type Item = Result<Bytes>;

    fn poll_next(
        self: Pin<&mut Self>,
        cx: &mut task::Context,
    ) -> task::Poll<Option<Result<Bytes>>> {
        let this = self.get_mut();

        if let Some(resp_body) = this.resp_body.take() {
            let chunk = encode_stream_item(&resp_body, this.encoding);
            return task::Poll::Ready(Some(Ok(chunk)));
        }

        match this.join_set.poll_join_next(cx) {
            task::Poll::Pending => {}
            task::Poll::Ready(Some(Ok(()))) => {}
            task::Poll::Ready(Some(Err(err))) => panic!("Cursor task crashed: {}", err),
            task::Poll::Ready(None) => {}
        };

        match this.cursor_hnd.poll_fetch(cx) {
            task::Poll::Pending => task::Poll::Pending,
            task::Poll::Ready(None) => task::Poll::Ready(None),
            task::Poll::Ready(Some(Ok(entry))) => {
                let chunk = encode_stream_item(&entry.entry, this.encoding);
                task::Poll::Ready(Some(Ok(chunk)))
            }
            task::Poll::Ready(Some(Err(err))) => task::Poll::Ready(Some(Err(err))),
        }
    }
}

fn encode_stream_item<T: Serialize + prost::Message>(item: &T, encoding: Encoding) -> Bytes {
    let mut data: Vec<u8>;
    match encoding {
        Encoding::Json => {
            data = serde_json::to_vec(item).unwrap();
            data.push(b'\n');
        }
        Encoding::Protobuf => {
            data = <T as prost::Message>::encode_length_delimited_to_vec(item);
        }
    }
    Bytes::from(data)
}

async fn read_decode_request<T: DeserializeOwned + prost::Message + Default>(
    req: hyper::Request<hyper::Body>,
    encoding: Encoding,
) -> Result<T> {
    let req_body = hyper::body::to_bytes(req.into_body())
        .await
        .context("Could not read request body")?;
    match encoding {
        Encoding::Json => serde_json::from_slice(&req_body)
            .map_err(|err| ProtocolError::JsonDeserialize { source: err })
            .context("Could not deserialize JSON request body"),
        Encoding::Protobuf => <T as prost::Message>::decode(req_body)
            .map_err(|err| ProtocolError::ProtobufDecode { source: err })
            .context("Could not decode Protobuf request body"),
    }
}

fn protocol_error_response(err: ProtocolError) -> hyper::Response<hyper::Body> {
    text_response(hyper::StatusCode::BAD_REQUEST, err.to_string())
}

fn stream_error_response(err: StreamError, encoding: Encoding) -> hyper::Response<hyper::Body> {
    let status = match err {
        StreamError::StreamExpired => hyper::StatusCode::BAD_REQUEST,
    };
    encode_response(
        status,
        &proto::Error {
            message: err.to_string(),
            code: err.code().into(),
        },
        encoding,
    )
}

fn encode_response<T: Serialize + prost::Message>(
    status: hyper::StatusCode,
    resp_body: &T,
    encoding: Encoding,
) -> hyper::Response<hyper::Body> {
    let (resp_body, content_type) = match encoding {
        Encoding::Json => (serde_json::to_vec(resp_body).unwrap(), "application/json"),
        Encoding::Protobuf => (
            <T as prost::Message>::encode_to_vec(resp_body),
            "application/x-protobuf",
        ),
    };
    hyper::Response::builder()
        .status(status)
        .header(hyper::http::header::CONTENT_TYPE, content_type)
        .body(hyper::Body::from(resp_body))
        .unwrap()
}

fn text_response(status: hyper::StatusCode, resp_body: String) -> hyper::Response<hyper::Body> {
    hyper::Response::builder()
        .status(status)
        .header(hyper::http::header::CONTENT_TYPE, "text/plain")
        .body(hyper::Body::from(resp_body))
        .unwrap()
}
