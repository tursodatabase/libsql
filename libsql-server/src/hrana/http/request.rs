use anyhow::{anyhow, bail, Result};
use bytesize::ByteSize;

use super::super::{batch, stmt, ProtocolError, Version};
use super::stream;
use crate::connection::{Connection, RequestContext};
use libsql_hrana::proto;

const MAX_SQL_COUNT: usize = 50;
const MAX_STORED_SQL_SIZE: ByteSize = ByteSize::kb(5);

/// An error from executing a [`proto::StreamRequest`]
#[derive(thiserror::Error, Debug)]
enum StreamResponseError {
    #[error("The server already stores {count} SQL texts, it cannot store more")]
    SqlTooMany { count: usize },
    #[error("The statement is too large to be stored: {size}/{MAX_STORED_SQL_SIZE}")]
    SqlTooLarge { size: ByteSize },
    #[error(transparent)]
    Stmt(stmt::StmtError),
    #[error(transparent)]
    Batch(batch::BatchError),
}

pub async fn handle(
    stream_guard: &mut stream::Guard<'_>,
    ctx: RequestContext,
    request: proto::StreamRequest,
    version: Version,
) -> Result<proto::StreamResult> {
    let result = match try_handle(stream_guard, ctx, request, version).await {
        Ok(response) => proto::StreamResult::Ok { response },
        Err(err) => {
            let resp_err = err.downcast::<StreamResponseError>()?;
            let error = proto::Error {
                message: resp_err.to_string(),
                code: resp_err.code().into(),
            };
            proto::StreamResult::Error { error }
        }
    };
    Ok(result)
}

async fn try_handle(
    stream_guard: &mut stream::Guard<'_>,
    ctx: RequestContext,
    request: proto::StreamRequest,
    version: Version,
) -> Result<proto::StreamResponse> {
    macro_rules! ensure_version {
        ($min_version:expr, $what:expr) => {
            if version < $min_version {
                bail!(ProtocolError::NotSupported {
                    what: $what,
                    min_version: $min_version,
                })
            }
        };
    }

    Ok(match request {
        proto::StreamRequest::None => bail!(ProtocolError::NoneStreamRequest),
        proto::StreamRequest::Close(_req) => {
            stream_guard.close_db();
            proto::StreamResponse::Close(proto::CloseStreamResp {})
        }
        proto::StreamRequest::Execute(req) => {
            let db = stream_guard.get_db()?;
            let sqls = stream_guard.sqls();
            let query =
                stmt::proto_stmt_to_query(&req.stmt, sqls, version).map_err(catch_stmt_error)?;
            let result = stmt::execute_stmt(db, ctx, query, req.stmt.replication_index)
                .await
                .map_err(catch_stmt_error)?;
            proto::StreamResponse::Execute(proto::ExecuteStreamResp { result })
        }
        proto::StreamRequest::Batch(req) => {
            let db = stream_guard.get_db()?;
            let sqls = stream_guard.sqls();
            let pgm = batch::proto_batch_to_program(&req.batch, sqls, version)
                .map_err(catch_stmt_error)
                .map_err(catch_batch_error)?;
            let result = batch::execute_batch(db, ctx, pgm, req.batch.replication_index)
                .await
                .map_err(catch_batch_error)?;
            proto::StreamResponse::Batch(proto::BatchStreamResp { result })
        }
        proto::StreamRequest::Sequence(req) => {
            let db = stream_guard.get_db()?;
            let sqls = stream_guard.sqls();
            let sql = stmt::proto_sql_to_sql(req.sql.as_deref(), req.sql_id, sqls, version)?;
            let pgm = batch::proto_sequence_to_program(sql).map_err(catch_stmt_error)?;
            batch::execute_sequence(db, ctx, pgm, req.replication_index)
                .await
                .map_err(catch_stmt_error)
                .map_err(catch_batch_error)?;
            proto::StreamResponse::Sequence(proto::SequenceStreamResp {})
        }
        proto::StreamRequest::Describe(req) => {
            let db = stream_guard.get_db()?;
            let sqls = stream_guard.sqls();
            let sql = stmt::proto_sql_to_sql(req.sql.as_deref(), req.sql_id, sqls, version)?;
            let result = stmt::describe_stmt(db, ctx, sql.into(), req.replication_index)
                .await
                .map_err(catch_stmt_error)?;
            proto::StreamResponse::Describe(proto::DescribeStreamResp { result })
        }
        proto::StreamRequest::StoreSql(req) => {
            let sqls = stream_guard.sqls_mut();
            let sql_id = req.sql_id;
            if sqls.contains_key(&sql_id) {
                bail!(ProtocolError::SqlExists { sql_id })
            } else if sqls.len() >= MAX_SQL_COUNT {
                bail!(StreamResponseError::SqlTooMany { count: sqls.len() })
            } else if req.sql.len() > MAX_STORED_SQL_SIZE.as_u64() as usize {
                bail!(StreamResponseError::SqlTooLarge {
                    size: ByteSize::b(req.sql.len() as _)
                })
            }
            sqls.insert(sql_id, req.sql);
            proto::StreamResponse::StoreSql(proto::StoreSqlStreamResp {})
        }
        proto::StreamRequest::CloseSql(req) => {
            let sqls = stream_guard.sqls_mut();
            sqls.remove(&req.sql_id);
            proto::StreamResponse::CloseSql(proto::CloseSqlStreamResp {})
        }
        proto::StreamRequest::GetAutocommit(_req) => {
            ensure_version!(Version::Hrana3, "The `get_autocommit` request");
            let db = stream_guard.get_db()?;
            let is_autocommit = db.is_autocommit().await?;
            proto::StreamResponse::GetAutocommit(proto::GetAutocommitStreamResp { is_autocommit })
        }
    })
}

fn catch_stmt_error(err: anyhow::Error) -> anyhow::Error {
    match err.downcast::<stmt::StmtError>() {
        Ok(stmt_err) => anyhow!(StreamResponseError::Stmt(stmt_err)),
        Err(err) => match err.downcast::<crate::Error>() {
            Ok(crate::Error::Migration(crate::schema::Error::MigrationError(_step, message))) => {
                anyhow!(StreamResponseError::Stmt(stmt::StmtError::SqliteError {
                    source: rusqlite::ffi::Error {
                        code: rusqlite::ffi::ErrorCode::Unknown,
                        extended_code: 4242
                    },
                    message
                }))
            }
            Ok(err) => anyhow!(err),
            Err(err) => err,
        },
    }
}

fn catch_batch_error(err: anyhow::Error) -> anyhow::Error {
    match err.downcast::<batch::BatchError>() {
        Ok(batch_err) => anyhow!(StreamResponseError::Batch(batch_err)),
        Err(err) => err,
    }
}

impl StreamResponseError {
    pub fn code(&self) -> &'static str {
        match self {
            Self::SqlTooMany { .. } => "SQL_STORE_TOO_MANY",
            Self::Stmt(err) => err.code(),
            Self::Batch(err) => err.code(),
            StreamResponseError::SqlTooLarge { .. } => "SQL_STORE_TOO_LARGE",
        }
    }
}
