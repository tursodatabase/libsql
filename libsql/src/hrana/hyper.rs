use crate::hrana::connection::HttpConnection;
use crate::hrana::pipeline::ServerMsg;
use crate::hrana::proto::Stmt;
use crate::hrana::{HranaError, HttpSend, Result};
use crate::params::Params;
use crate::util::ConnectorService;
use crate::{Rows, Statement};
use futures::future::BoxFuture;
use http::header::AUTHORIZATION;
use http::{HeaderValue, StatusCode};
use hyper_rustls::{HttpsConnector, HttpsConnectorBuilder};

#[derive(Clone, Debug)]
pub struct HttpSender {
    inner: hyper::Client<HttpsConnector<ConnectorService>, hyper::Body>,
    version: HeaderValue,
}

impl HttpSender {
    pub fn new(connector: ConnectorService, version: Option<&str>) -> Self {
        let ver = version.unwrap_or_else(|| env!("CARGO_PKG_VERSION"));

        let version = HeaderValue::try_from(format!("libsql-remote-{ver}")).unwrap();

        let https = HttpsConnectorBuilder::new()
            .with_native_roots()
            .https_or_http()
            .enable_http1()
            .wrap_connector(connector);
        let inner = hyper::Client::builder().build(https);

        Self { inner, version }
    }

    async fn send(&self, url: String, auth: String, body: String) -> Result<ServerMsg> {
        let req = hyper::Request::post(url)
            .header(AUTHORIZATION, auth)
            .header("x-libsql-client-version", self.version.clone())
            .body(hyper::Body::from(body))
            .unwrap();

        let res = self.inner.request(req).await.map_err(HranaError::from)?;

        if res.status() != StatusCode::OK {
            let body = hyper::body::to_bytes(res.into_body())
                .await
                .map_err(HranaError::from)?;
            let body = String::from_utf8(body.into()).unwrap();
            return Err(HranaError::Api(body));
        }

        let body = hyper::body::to_bytes(res.into_body())
            .await
            .map_err(HranaError::from)?;

        let msg = serde_json::from_slice::<ServerMsg>(&body[..]).map_err(HranaError::from)?;

        Ok(msg)
    }
}

impl<'a> HttpSend<'a> for HttpSender {
    type Result = BoxFuture<'a, Result<ServerMsg>>;

    fn http_send(&'a self, url: String, auth: String, body: String) -> Self::Result {
        let fut = self.send(url, auth, body);
        Box::pin(fut)
    }
}

impl From<hyper::Error> for HranaError {
    fn from(value: hyper::Error) -> Self {
        HranaError::Http(value.to_string())
    }
}

impl HttpConnection<HttpSender> {
    pub(crate) fn new_with_connector(
        url: impl Into<String>,
        token: impl Into<String>,
        connector: ConnectorService,
        version: Option<&str>,
    ) -> Self {
        let inner = HttpSender::new(connector, version);
        Self::new(url.into(), token.into(), inner)
    }
}

#[async_trait::async_trait]
impl crate::connection::Conn for HttpConnection<HttpSender> {
    async fn execute(&self, sql: &str, params: Params) -> crate::Result<u64> {
        let mut stmt = self.prepare(sql);
        let rows = stmt.execute(&params).await?;

        Ok(rows as u64)
    }

    async fn execute_batch(&self, sql: &str) -> crate::Result<()> {
        let mut statements = Vec::new();
        let stmts = crate::parser::Statement::parse(sql);
        for s in stmts {
            let s = s?;
            statements.push(Stmt::new(s.stmt, false));
        }
        self.raw_batch(statements)
            .await
            .map_err(|e| crate::Error::Hrana(e.into()))?;
        Ok(())
    }

    async fn prepare(&self, sql: &str) -> crate::Result<Statement> {
        let stmt = crate::hrana::Statement::new(self.clone(), sql.to_string(), true);
        Ok(Statement {
            inner: Box::new(stmt),
        })
    }

    async fn transaction(
        &self,
        _tx_behavior: crate::TransactionBehavior,
    ) -> crate::Result<crate::transaction::Transaction> {
        todo!()
    }

    fn is_autocommit(&self) -> bool {
        // TODO: Is this correct?
        false
    }

    fn changes(&self) -> u64 {
        self.affected_row_count()
    }

    fn last_insert_rowid(&self) -> i64 {
        self.last_insert_rowid()
    }

    fn close(&mut self) {
        todo!()
    }
}

#[async_trait::async_trait]
impl crate::statement::Stmt for crate::hrana::Statement<HttpSender> {
    fn finalize(&mut self) {}

    async fn execute(&mut self, params: &Params) -> crate::Result<usize> {
        self.execute(params).await
    }

    async fn query(&mut self, params: &Params) -> crate::Result<Rows> {
        self.query(params).await
    }

    fn reset(&mut self) {}

    fn parameter_count(&self) -> usize {
        todo!()
    }

    fn parameter_name(&self, _idx: i32) -> Option<&str> {
        todo!()
    }

    fn columns(&self) -> Vec<crate::Column> {
        todo!()
    }
}
