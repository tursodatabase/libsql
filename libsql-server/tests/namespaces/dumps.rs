use std::convert::Infallible;
use std::time::Duration;

use hyper::{service::make_service_fn, Body, Response, StatusCode};
use insta::{assert_json_snapshot, assert_snapshot};
use libsql::{Database, Value};
use serde_json::json;
use tempfile::tempdir;
use tower::service_fn;
use turmoil::Builder;

use crate::common::http::Client;
use crate::common::net::{TurmoilAcceptor, TurmoilConnector};
use crate::namespaces::make_primary;

#[test]
fn load_namespace_from_dump_from_url() {
    const DUMP: &str = r#"
        PRAGMA foreign_keys=OFF;
    BEGIN TRANSACTION;
    CREATE TABLE test (x);
    INSERT INTO test VALUES(42);
    COMMIT;"#;

    let mut sim = Builder::new()
        .simulation_duration(Duration::from_secs(1000))
        .build();
    let tmp = tempdir().unwrap();
    make_primary(&mut sim, tmp.path().to_path_buf());

    sim.host("dump-store", || async {
        let incoming = TurmoilAcceptor::bind(([0, 0, 0, 0], 8080)).await?;
        let server =
            hyper::server::Server::builder(incoming).serve(make_service_fn(|_conn| async {
                Ok::<_, Infallible>(service_fn(|_req| async {
                    Ok::<_, Infallible>(Response::new(Body::from(DUMP)))
                }))
            }));

        server.await.unwrap();

        Ok(())
    });

    sim.client("client", async {
        let client = Client::new();
        let resp = client
            .post(
                "http://primary:9090/v1/namespaces/foo/create",
                json!({ "dump_url": "http://dump-store:8080/"}),
            )
            .await
            .unwrap();
        assert_eq!(resp.status(), 200);
        assert_snapshot!(resp.body_string().await.unwrap());

        let foo =
            Database::open_remote_with_connector("http://foo.primary:8080", "", TurmoilConnector)?;
        let foo_conn = foo.connect()?;
        let mut rows = foo_conn.query("select count(*) from test", ()).await?;
        assert!(matches!(
            rows.next().await.unwrap().unwrap().get_value(0)?,
            Value::Integer(1)
        ));

        Ok(())
    });

    sim.run().unwrap();
}

#[test]
fn load_namespace_from_dump_from_file() {
    const DUMP: &str = r#"
        PRAGMA foreign_keys=OFF;
    BEGIN TRANSACTION;
    CREATE TABLE test (x);
    INSERT INTO test VALUES(42);
    COMMIT;"#;

    let mut sim = Builder::new()
        .simulation_duration(Duration::from_secs(1000))
        .build();
    let tmp = tempdir().unwrap();
    let tmp_path = tmp.path().to_path_buf();

    std::fs::write(tmp_path.join("dump.sql"), DUMP).unwrap();

    make_primary(&mut sim, tmp.path().to_path_buf());

    sim.client("client", async move {
        let client = Client::new();

        // path is not absolute is an error
        let resp = client
            .post(
                "http://primary:9090/v1/namespaces/foo/create",
                json!({ "dump_url": "file:dump.sql"}),
            )
            .await
            .unwrap();
        assert_eq!(resp.status(), StatusCode::BAD_REQUEST);

        // path doesn't exist is an error
        let resp = client
            .post(
                "http://primary:9090/v1/namespaces/foo/create",
                json!({ "dump_url": "file:/dump.sql"}),
            )
            .await
            .unwrap();
        assert_eq!(resp.status(), StatusCode::BAD_REQUEST);

        let resp = client
            .post(
                "http://primary:9090/v1/namespaces/foo/create",
                json!({ "dump_url": format!("file:{}", tmp_path.join("dump.sql").display())}),
            )
            .await
            .unwrap();
        assert_eq!(
            resp.status(),
            StatusCode::OK,
            "{}",
            resp.json::<serde_json::Value>().await.unwrap_or_default()
        );

        let foo =
            Database::open_remote_with_connector("http://foo.primary:8080", "", TurmoilConnector)?;
        let foo_conn = foo.connect()?;
        let mut rows = foo_conn.query("select count(*) from test", ()).await?;
        assert!(matches!(
            rows.next().await.unwrap().unwrap().get_value(0)?,
            Value::Integer(1)
        ));

        Ok(())
    });

    sim.run().unwrap();
}

#[test]
fn load_namespace_from_no_commit() {
    const DUMP: &str = r#"
    PRAGMA foreign_keys=OFF;
    BEGIN TRANSACTION;
    CREATE TABLE test (x);
    INSERT INTO test VALUES(42);
    "#;

    let mut sim = Builder::new()
        .simulation_duration(Duration::from_secs(1000))
        .build();
    let tmp = tempdir().unwrap();
    let tmp_path = tmp.path().to_path_buf();

    std::fs::write(tmp_path.join("dump.sql"), DUMP).unwrap();

    make_primary(&mut sim, tmp.path().to_path_buf());

    sim.client("client", async move {
        let client = Client::new();
        let resp = client
            .post(
                "http://primary:9090/v1/namespaces/foo/create",
                json!({ "dump_url": format!("file:{}", tmp_path.join("dump.sql").display())}),
            )
            .await
            .unwrap();
        // the dump is malformed
        assert_eq!(
            resp.status(),
            StatusCode::BAD_REQUEST,
            "{}",
            resp.json::<serde_json::Value>().await.unwrap_or_default()
        );

        // namespace doesn't exist
        let foo =
            Database::open_remote_with_connector("http://foo.primary:8080", "", TurmoilConnector)?;
        let foo_conn = foo.connect()?;
        assert!(foo_conn
            .query("select count(*) from test", ())
            .await
            .is_err());

        Ok(())
    });

    sim.run().unwrap();
}

#[test]
fn load_namespace_from_no_txn() {
    const DUMP: &str = r#"
    PRAGMA foreign_keys=OFF;
    CREATE TABLE test (x);
    INSERT INTO test VALUES(42);
    COMMIT;
    "#;

    let mut sim = Builder::new()
        .simulation_duration(Duration::from_secs(1000))
        .build();
    let tmp = tempdir().unwrap();
    let tmp_path = tmp.path().to_path_buf();

    std::fs::write(tmp_path.join("dump.sql"), DUMP).unwrap();

    make_primary(&mut sim, tmp.path().to_path_buf());

    sim.client("client", async move {
        let client = Client::new();
        let resp = client
            .post(
                "http://primary:9090/v1/namespaces/foo/create",
                json!({ "dump_url": format!("file:{}", tmp_path.join("dump.sql").display())}),
            )
            .await?;
        // the dump is malformed
        assert_eq!(
            resp.status(),
            StatusCode::BAD_REQUEST,
            "{}",
            resp.json::<serde_json::Value>().await.unwrap_or_default()
        );
        assert_json_snapshot!(resp.json_value().await.unwrap());

        // namespace doesn't exist
        let foo =
            Database::open_remote_with_connector("http://foo.primary:8080", "", TurmoilConnector)?;
        let foo_conn = foo.connect()?;
        assert!(foo_conn
            .query("select count(*) from test", ())
            .await
            .is_err());

        Ok(())
    });

    sim.run().unwrap();
}

#[test]
fn export_dump() {
    let mut sim = Builder::new()
        .simulation_duration(Duration::from_secs(1000))
        .build();
    let tmp = tempdir().unwrap();

    make_primary(&mut sim, tmp.path().to_path_buf());

    sim.client("client", async move {
        let client = Client::new();
        let resp = client
            .post("http://primary:9090/v1/namespaces/foo/create", json!({}))
            .await?;
        assert_eq!(resp.status(), StatusCode::OK);

        let foo =
            Database::open_remote_with_connector("http://foo.primary:8080", "", TurmoilConnector)?;
        let foo_conn = foo.connect()?;
        foo_conn.execute("create table test (x)", ()).await?;
        foo_conn.execute("insert into test values (42)", ()).await?;
        foo_conn
            .execute("insert into test values ('foo')", ())
            .await?;
        foo_conn
            .execute("insert into test values ('bar')", ())
            .await?;

        let resp = client.get("http://foo.primary:8080/dump").await?;
        assert_eq!(resp.status(), StatusCode::OK);
        assert_snapshot!(resp.body_string().await?);

        Ok(())
    });

    sim.run().unwrap();
}
