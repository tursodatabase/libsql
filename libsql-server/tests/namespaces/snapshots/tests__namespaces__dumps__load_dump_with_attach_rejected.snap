---
source: libsql-server/tests/namespaces/dumps.rs
expression: resp.body_string().await?
snapshot_kind: text
---
{"error":"The passed dump sql is invalid: attach statements are not allowed in dumps, msg: near \"COMMIT\": syntax error, sql:     ATTACH foo/bar.sql\n     COMMIT;, offset: 28"}
